### PR TITLE
GH-3296: Add objectStorageConfigFile for ThanosSpec

### PR DESCRIPTION
**Pull request by TamasNeumer**
Hi!,

I have added the option to specify the objectStorageCofnigFile, as desired by the [Thanos Sidecar Specification](https://github.com/thanos-io/thanos/blob/master/docs/components/sidecar.md#flags).

I did the changes for both sidecar and ruler.

I added two tests (sidecar, ruler), testing the desired behavior. 

@brancz, looks okay? Do you guys have gitter or discord channel?

I'm not sure if I understand the following part:

```yaml
$GOPATH/src/github.com/prometheus-operator/prometheus-operator as tests expect paths to match. 
If you're working from a fork, just add the forked repo as a remote and pull against your local 
prometheus-operator checkout before running tests.
```

Thanks!

### CODE REVIEW

**Comment by brancz** [comment]
Looks like some CI jobs are failing. For the generate job, you can try running `make clean generate` and commit the changes.

There's a development channel called #prometheus-operator-dev on kubernetes slack.

**Comment by TamasNeumer** [comment]
@brancz, I have fixed the test. Running `make clean generate` didn't help (no files were generated, which were under version control.)

What should/can I do to make it pass?
It's a simple feature, but we would need it quite soon.

thanks!

File: Documentation/api.md [hunk 1]
```diff
@@ -850,6 +850,7 @@ ThanosSpec defines parameters for a Prometheus server within a Thanos deployment
 | baseImage | Thanos base image if other than default. Deprecated: use 'image' instead | *string | false |
 | resources | Resources defines the resource requirements for the Thanos sidecar. If not provided, no requests/limits will be set | [v1.ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#resourcerequirements-v1-core) | false |
 | objectStorageConfig | ObjectStorageConfig configures object storage in Thanos. | *[v1.SecretKeySelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#secretkeyselector-v1-core) | false |
+| ObjectStorageConfigFile | ObjectStorageConfigFile specifies the path of the object storage configuration file. Can't be used with ObjectStorageConfig at the same time. | *string | false |
```
Comment (lilic) [inline Documentation/api.md:853]: As mentioned on slack this should have a lower case start for this field, so camelCase not PascalCase.

**Review by lilic — COMMENTED** [review]
Thanks for the PR. One initial comment, will have a more in depth look later.

File: Documentation/api.md (hunk 1 above)
Comment (TamasNeumer) [reply Documentation/api.md]: Thanks!

Fixed and pushed!

File: pkg/prometheus/statefulset.go [hunk 2]
```diff
@@ -749,14 +750,18 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
...
+		if p.Spec.Thanos.ObjectStorageConfig != nil || p.Spec.Thanos.ObjectStorageConfigFile != nil {
+			if p.Spec.Thanos.ObjectStorageConfig != nil {
+				container.Args = append(container.Args, "--objstore.config=$(OBJSTORE_CONFIG)")
+				container.Env = append(container.Env, v1.EnvVar{
+					Name: "OBJSTORE_CONFIG",
+					ValueFrom: &v1.EnvVarSource{
+						SecretKeyRef: p.Spec.Thanos.ObjectStorageConfig,
+					},
+				})
+			} else {
```
Comment (lilic) [inline pkg/prometheus/statefulset.go:762]: I would not do the else here and instead keep what we had before and just do something like:

```
if p.Spec.Thanos.ObjectStorageConfigFile != nil {
// append the container args here
}
```

This was we let thanos decide the ordering and if user sets both, both are passed.

**Review by lilic — COMMENTED** [review]
Thanks! Few comments and suggestions.

File: pkg/apis/monitoring/v1/thanos_types.go [hunk 3]
```diff
@@ -96,6 +96,9 @@ type ThanosRulerSpec struct {
 	Volumes []v1.Volume `json:"volumes,omitempty"`
 	// ObjectStorageConfig configures object storage in Thanos.
```
[the remark this answers is not in the document]
Comment (TamasNeumer) [reply pkg/apis/monitoring/v1/thanos_types.go:97]: I have absolutely no idea how thanos behaves if both flags are passed. (Which is lower/higher order).

[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/apis/monitoring/v1/thanos_types.go:97]: What @lilic says, see the [Thanos docs](https://thanos.io/v0.16/components/sidecar.md/#flags), `ObjectStorageConfigFile` :)

[the remark this answers is not in the document]
Comment (TamasNeumer) [reply pkg/apis/monitoring/v1/thanos_types.go:97]: Thanks. Makes sense. I was looking at [here](https://thanos.io/v0.16/components/sidecar.md/), but there they don't mention it. Thanks for the clarification! :-)

**Review by lilic — COMMENTED** [review]
Nice! lgtm giving @simonpasquier a chance to have a look as well.

**Review by simonpasquier — APPROVED** [review]

**Comment by simonpasquier** [comment]
Thanks!

**Pull request merged**